File: 09. Graph/09. Kruskal's MST Algorithm/Kruskals.py
class DisjointSet:
    def __init__(self, n):
        self.parent = [i for i in range(n)]  # Each element is its own parent
        self.rank = [0] * n  # Rank is initially 0
        self.size = [1] * n  # Size of each set is initially 1

    # Find the representative (root) of the set containing x, with path compression
    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])  # Path compression
        return self.parent[x]

    # Union two sets containing x and y, using union by rank
    def unionByRank(self, x, y):
        rootX = self.find(x)
        rootY = self.find(y)

        if rootX != rootY:
            if self.rank[rootX] > self.rank[rootY]:
                self.parent[rootY] = rootX  # Attach smaller tree under larger tree
            elif self.rank[rootX] < self.rank[rootY]:
                self.parent[rootX] = rootY
            else:
                self.parent[rootY] = rootX  # If ranks are equal, attach under one and increment rank
                self.rank[rootX] += 1

    # Union two sets containing x and y, using union by size
    def unionBySize(self, x, y):
        rootX = self.find(x)
        rootY = self.find(y)

        if rootX != rootY:
            if self.size[rootX] > self.size[rootY]:
                self.parent[rootY] = rootX
                self.size[rootX] += self.size[rootY]
            else:
                self.parent[rootX] = rootY
                self.size[rootY] += self.size[rootX]

    # Helper function to check if two elements are in the same set
    def is_connected(self, x, y):
        return self.find(x) == self.find(y)

class Edge:
    def __init__(self, u, v, weight):
        self.u = u
        self.v = v
        self.weight = weight

    # Comparator to sort edges by weight
    def __lt__(self, other):
        return self.weight < other.weight

def kruskalMST(n, edges):
    # Sort edges by weight
    edges.sort()

    ds = DisjointSet(n)

    mst = []
    total_weight = 0

    for edge in edges:
        u = edge.u
        v = edge.v

        # If they are in different sets, include this edge in MST
        if ds.find(u) != ds.find(v):
            ds.unionByRank(u, v)
            mst.append(edge)
            total_weight += edge.weight

    # Output the MST
    print("Edges in the Minimum Spanning Tree:")
    for edge in mst:
        print(f"{edge.u} - {edge.v} (weight: {edge.weight})")
    print(f"Total weight of MST: {total_weight}")


def main():
    # Number of vertices
    n = 5
    edges = [
        Edge(0, 1, 2),
        Edge(0, 3, 6),
        Edge(1, 2, 3),
        Edge(1, 3, 8),
        Edge(1, 4, 5),
        Edge(2, 4, 7)
    ]

    # Call Kruskal's Algorithm
    kruskalMST(n, edges)

main()

File: 09. Graph/09. Kruskal's MST Algorithm/Kruskals.go
package main

import (
	"fmt"
	"sort"
)

// ---------- Disjoint Set (Union-Find) ----------

type DisjointSet struct {
	parent []int
	rank   []int
}

func NewDisjointSet(n int) *DisjointSet {
	ds := &DisjointSet{
		parent: make([]int, n),
		rank:   make([]int, n),
	}
	for i := 0; i < n; i++ {
		ds.parent[i] = i
	}
	return ds
}

func (ds *DisjointSet) Find(x int) int {
	if ds.parent[x] != x {
		ds.parent[x] = ds.Find(ds.parent[x]) // Path compression
	}
	return ds.parent[x]
}

func (ds *DisjointSet) UnionByRank(x, y int) {
	rootX := ds.Find(x)
	rootY := ds.Find(y)

	if rootX != rootY {
		if ds.rank[rootX] > ds.rank[rootY] {
			ds.parent[rootY] = rootX
		} else if ds.rank[rootX] < ds.rank[rootY] {
			ds.parent[rootX] = rootY
		} else {
			ds.parent[rootY] = rootX
			ds.rank[rootX]++
		}
	}
}

// ---------- Edge Structure ----------

type Edge struct {
	u, v, weight int
}

// ---------- Kruskal's Algorithm ----------

func kruskalMST(n int, edges []Edge) {
	// Sort edges by weight
	sort.Slice(edges, func(i, j int) bool {
		return edges[i].weight < edges[j].weight
	})

	ds := NewDisjointSet(n)
	totalWeight := 0
	mst := []Edge{}

	for _, edge := range edges {
		if ds.Find(edge.u) != ds.Find(edge.v) {
			ds.UnionByRank(edge.u, edge.v)
			mst = append(mst, edge)
			totalWeight += edge.weight
		}
	}

	// Print MST
	fmt.Println("Edges in the Minimum Spanning Tree:")
	for _, edge := range mst {
		fmt.Printf("%d - %d (weight: %d)\n", edge.u, edge.v, edge.weight)
	}
	fmt.Println("Total weight of MST:", totalWeight)
}

// ---------- Main ----------

func main() {
	n := 5

	edges := []Edge{
		{0, 1, 2},
		{0, 3, 6},
		{1, 2, 3},
		{1, 3, 8},
		{1, 4, 5},
		{2, 4, 7},
	}

	kruskalMST(n, edges)
}